File: C++_ver/hdr.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include<tuple>

#include<time.h>

#include "hdr.h"

void float2rgbe(unsigned char rgbe[4], float red, float green, float blue) {
	float v;
	int e;
	v = red;
	if (green > v) {
		v = green;
	}
	if (blue > v) {
		v = blue;
	}
	if (v < 1e-32) {
		rgbe[0] = rgbe[1] = rgbe[2] = rgbe[3] = 0;
	}
	else {
		v = frexp(v, &e) * 256 / v;
		rgbe[0] = (unsigned int)(red * v);
		rgbe[1] = (unsigned int)(green * v);
		rgbe[2] = (unsigned int)(blue * v);
		rgbe[3] = (unsigned int)(e + 128);
	}
}

void HDRwriter(std::string filename, cv::Mat img) {
	std::ofstream hdr(filename, std::ios::binary);
	hdr << "#?RADIANCE\n";
	hdr << "FORMAT=32-bit_rle_rgbe\n\n";
	hdr << "-Y " << img.rows << " +X " << img.cols << "\n";
	unsigned char rgbe[4];
	for (int i = 0; i < img.rows; i++) {
		for (int j = 0; j < img.cols; j++) {
			float2rgbe(rgbe, img.at<cv::Vec3f>(i, j)[2], img.at<cv::Vec3f>(i, j)[1], img.at<cv::Vec3f>(i, j)[0]);
			hdr.write(reinterpret_cast<const char*> (rgbe), sizeof(rgbe));
		}
	}
	hdr.close();
}

//reference from debevec's HDR method (gsolve.m)
std::tuple <cv::Mat, cv::Mat> gsolve(std::vector<std::vector<int>>& Z, std::vector<double>& B, int l, std::vector<int>& w) {
	int n = 256;
	cv::Mat A = cv::Mat::zeros(Z.size() * Z[0].size() + n + 1, n + Z.size(), CV_64F);
	cv::Mat b = cv::Mat::zeros(A.rows, 1, CV_64F);


	int k = 0;
	for (int i = 0; i < Z.size(); i++) {
		for (int j = 0; j < Z[0].size(); j++) {
			double wij = w[Z[i][j]];
			//::cout << wij * B[j];
			A.at<double>(k, Z[i][j]) = wij;
			A.at<double>(k, n + i) = -wij;
			b.at<double>(k, 0) = wij * B[j];
			k++;

		}
	}

	A.at<double>(k, 128) = 1;
	k++;

	for (int i = 0; i < n - 2; i++) {
		A.at<double>(k, i) = (double)l * w[i + 1];
		A.at<double>(k, i + 1) = (double)-2 * l * w[i + 1];
		A.at<double>(k, i + 2) = (double)l * w[i + 1];
		k++;

	}

	cv::Mat x;
	cv::solve(A, b, x, cv::DECOMP_SVD);



	cv::Mat g = x(cv::Range(0, n), cv::Range(0, x.cols));
	cv::Mat lE = x(cv::Range(n, x.rows), cv::Range(0, x.cols));

	return { g, lE };
}

void HDR(std::vector<std::string>& Imglist, std::vector<double>& Explist) {

	clock_t start = clock();

	std::vector<cv::Mat> cImgs;
	for (std::string img : Imglist) {
		cImgs.push_back(cv::imread("Aligned_" + img, cv::IMREAD_UNCHANGED));
	}

	// sample 100 points
	int SampleSize = 100;
	std::vector<std::vector<std::vector<int>>> Z;
	for (int ch = 0; ch < 3; ch++) {
		Z.push_back(std::vector<std::vector<int>>());
		for (int i = 0; i < SampleSize; i++) {
			int x = rand() % cImgs[0].rows;
			int y = rand() % cImgs[0].cols;
			Z[ch].push_back(std::vector<int>());
			for (int j = 0; j < cImgs.size(); j++) {
				Z[ch][i].push_back(cImgs[j].at<cv::Vec3b>(x, y)[ch]);
			}
		}
	}
	//Read exposure
	std::vector<double> B(Explist.size(), 0);
	for (int ex = 0; ex < Explist.size(); ex++) {
		B[ex] = log(Explist[ex]);
	}

	//Calculate weighting function w(z)
	std::vector<int> w(256);
	for (int i = 0; i < 256; i++) {
		w[i] = (i <= 127) ? i : 255 - i;
	}
	//lamda l
	int l = 50;
	std::vector<cv::Mat> lElist;
	std::vector<cv::Mat> glist;

	for (int ch = 0; ch < 3; ch++) {
		cv::Mat g, lE;
		std::tie(g, lE) = gsolve(Z[ch], B, l, w);
		glist.push_back(g);
		lElist.push_back(lE);
	}

	cv::Mat E = cv::Mat::zeros(cImgs[0].rows, cImgs[0].cols, CV_32FC3);
	for (int ch = 0; ch < 3; ch++) {
		for (int i = 0; i < cImgs[0].rows; i++) {
			for (int j = 0; j < cImgs[0].cols; j++) {
				double w_sum = 1e-8;
				double ln_rad_sum = 0;
				for (int ind = 0; ind < cImgs.size(); ind++) {
					int pix = cImgs[ind].at<cv::Vec3b>(i, j)[ch];
					double w_pix = w[pix];
					double g_pix = glist[ch].at<double>(pix, 0);
					ln_rad_sum += w_pix * (glist[ch].at<double>(pix, 0) - B[ind]);
					w_sum += w_pix;

				}

				E.at<cv::Vec3f>(i, j)[ch] = (float)exp(ln_rad_sum / w_sum);
			}
		}
	}


	clock_t end = clock();

	std::cout << "Time of solving HDR: " << difftime(end, start) <<std::endl;

	HDRwriter("HDR.hdr", E);

}

File: python_ver/HDR.py
import cv2
import numpy as np
from random import randint
import matplotlib.pyplot as plt
import math

from mpl_toolkits.axes_grid1 import make_axes_locatable






#reference from debevec's HDR method (gsolve.m)
def gsolve(Z, B, l, w):
    n = 256
    A = np.zeros((Z.shape[0] * Z.shape[1] + n + 1, n + Z.shape[0]))
    b = np.zeros((A.shape[0],1))

    #Include the data-fitting equations
    k = 0
    for i in range(Z.shape[0]):
        for j in range(Z.shape[1]):
            wij = w[Z[i, j] ]
            A[k, Z[i, j]] = wij
            A[k, n+i] = -wij
            b[k, 0] = wij * B[j]
            k = k+1

    A[k, 128] = 1
    k = k+1

    #include the smoothness equations
    for i in range (n-2):
        A[k, i] = l * w[i+1]
        A[k, i+1] = -2 * l * w[i+1]
        A[k, i+2] = l * w[i+1]
        k= k+1
    #solve the system using SVD
    x = np.linalg.lstsq(A, b)[0]

    g = x[0:n]
    lE = x[n:x.shape[0]]

    return g, lE

def PlotResponseCurve(gList):
    plt.figure(figsize=(8, 8))
    blueline, = plt.plot(gList[0], range(256),color = 'blue')
    greenline, = plt.plot(gList[1], range(256),color = 'green')
    redline, = plt.plot(gList[2], range(256),color = 'red')
    plt.suptitle('Response Curve')
    plt.ylabel('Pixel Value (Zij)')
    plt.xlabel('Log Exposure (Ei*(delta t)j)')
    plt.legend(handles=[blueline, greenline, redline], labels = ['Blue', 'Green', 'Red'], loc = 'best', fontsize = 'x-large')
    plt.savefig('RC.png')


def PlotRadianceMap(lnE, filename):
    fig = plt.figure(figsize=(12, 5))
    ax1,ax2, ax3 = fig.subplots(1, 3)
    for ax, color, index in zip((ax1, ax2, ax3), ('Blue', 'Green', 'Red'), (0,1,2)):
        im = ax.imshow(lnE[:,:,index], cmap='jet')
        ax.set_axis_off()
        divider = make_axes_locatable(ax)
        cax = divider.append_axes("right", size="7%", pad="2%")
        ax.set_title(color, fontsize = 20)
        cbar = fig.colorbar(im, cax = cax)
        cbar.ax.tick_params(labelsize=8)
    fig.savefig(filename)


#Numpy radiance writer
def RadianceSave(HDR, filename):
    image = np.zeros(HDR.shape)
    image[:,:,0] = HDR[:,:,2]
    image[:,:,1] = HDR[:,:,1]
    image[:,:,2] = HDR[:,:,0]
    f = open(filename + ".hdr", "wb")
    f.write(b"#?RADIANCE\n# Made with Python & Numpy\nFORMAT=32-bit_rle_rgbe\n\n")
    title = ("-Y {0} +X {1}\n".format(image.shape[0], image.shape[1]))
    f.write(title.encode(encoding='UTF-8'))

    brightest = np.maximum(np.maximum(image[...,0], image[...,1]), image[...,2])
    mantissa = np.zeros_like(brightest)
    exponent = np.zeros_like(brightest)
    np.frexp(brightest, mantissa, exponent)

    ## 0-255 , 256 will have noisy in the result image
    scaled_mantissa = mantissa * 255.0 / brightest
    rgbe = np.zeros((image.shape[0], image.shape[1], 4), dtype=np.uint8)
    rgbe[...,0:3] = np.around(image[...,0:3] * scaled_mantissa[...,None])
    rgbe[...,3] = np.around(exponent + 128)

    rgbe.flatten().tofile(f)
    f.close()

      
def HDR(ImgList, ExposureList):
    #Read Image
    cImgs = [cv2.imread("Aligned_"+ImgPath, 1) for ImgPath in ImgList]    


    #Samping 100 points
    SampleSize = 100
    Row = cImgs[0].shape[0]
    Col = cImgs[0].shape[1]
    Channels = cImgs[0].shape[2]

    Z = np.zeros((SampleSize, len(cImgs), 3), dtype = int)
    for i in range(SampleSize):
        y = randint(0, Row - 1)
        x = randint(0, Col - 1)
        for j in range(len(cImgs)):
            Z[i][j] = cImgs[j][y][x]

    #Read exposure 
    ExpList = [float(i) for i in ExposureList]
    B = np.log(ExpList)

    #Calculate weighting function w(z)
    w = [z if z <= 0.5*255 else 255 - z for z in range(256)]

    #lamda l : smooth factor
    l = 50


    gList = []
    lEList = []
    for channel in range (Channels):
        Z_channel = Z[:,:,channel]
        g, lE = gsolve(Z_channel, B, l, w)
        gList.append(g)
        lEList.append(lE)


    #Constructing HDR rafiance map
    
    E = np.zeros(cImgs[0].shape)
    #w nparray for indices 
    w = np.array(w)
    for channel in range(Channels):
        w_sum = np.zeros((Row, Col))
        ln_rad_sum = np.zeros((Row, Col))
        for index, img in enumerate(cImgs):
            pix = img[:,:,channel].flatten()
            w_pix = w[pix].reshape(Row, Col)
            g_pix = (gList[channel][pix] - B[index]).reshape(Row, Col)
            ln_rad_sum += w_pix * g_pix
            w_sum += w_pix

        E[:,:,channel] = np.exp(ln_rad_sum / (w_sum + 1e-8))
    print(np.log(E).max())

    PlotRadianceMap(np.log(E), 'RadianceMap.png')

    PlotResponseCurve(gList)
    RadianceSave(E, "HDR")







if __name__ == '__main__':
    with open('image_list.txt', 'r') as f:
        ImgList = [line.strip() for line in f]
    with open('exposure.txt', 'r') as f:
        ExposureList = [line.strip() for line in f]

    HDR(ImgList, ExposureList)
